core/lib/web3_decl/src/error.rs

```diff
@@ -42,3 +55,152 @@ pub enum Web3Error {
     #[error("Tree API is not available")]
     TreeApiUnavailable,
 }
+
+/// Client RPC error with additional details: the method name and arguments of the called method.
+///
+/// The wrapped error can be accessed using [`AsRef`].
+#[derive(Debug)]
+pub struct EnrichedClientError {
+    inner_error: ClientError,
+    method: &'static str,
+    args: HashMap<&'static str, String>,
+}
+
+/// Alias for a result with enriched client RPC error.
+pub type EnrichedClientResult<T> = Result<T, EnrichedClientError>;
+
+impl EnrichedClientError {
+    /// Wraps the specified `inner_error`.
+    pub fn new(inner_error: ClientError, method: &'static str) -> Self {
+        Self {
+            inner_error,
+            method,
+            args: HashMap::new(),
+        }
+    }
+
+    /// Creates an error wrapping [`RpcError::Custom`].
+    pub fn custom(message: impl Into<String>, method: &'static str) -> Self {
+        Self::new(ClientError::Custom(message.into()), method)
+    }
+
+    /// Adds a tracked argument for this error.
+    #[must_use]
+    pub fn with_arg(mut self, name: &'static str, value: &dyn fmt::Debug) -> Self {
+        self.args.insert(name, format!("{value:?}"));
+        self
+    }
+}
+
+impl AsRef<ClientError> for EnrichedClientError {
+    fn as_ref(&self) -> &ClientError {
+        &self.inner_error
+    }
+}
+
+impl error::Error for EnrichedClientError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        Some(&self.inner_error)
+    }
+}
+
+impl fmt::Display for EnrichedClientError {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        struct DebugArgs<'a>(&'a HashMap<&'static str, String>);
+
+        impl fmt::Debug for DebugArgs<'_> {
+            fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                formatter.write_str("(")?;
+                for (i, (name, value)) in self.0.iter().enumerate() {
+                    write!(formatter, "{name}={value}")?;
+                    if i + 1 < self.0.len() {
+                        formatter.write_str(", ")?;
+                    }
+                }
+                formatter.write_str(")")
```

Comment:
just curious - could we do something like `.join(", ")` here? Did you want to avoid allocating a vector for this?